File: Arrays/P027.py
def brute(grid):
    n=len(grid)
    res=[[0]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            res[j][n-1-i]=grid[i][j]
    return res

def optimal(grid):
    n=len(grid)
    for i in range(n):
        for j in range(i+1,n):
            grid[i][j],grid[j][i]=grid[j][i],grid[i][j]
    for i in range(n):
        grid[i].reverse()
    return grid

mat2 = [
  [1, 2, 3],
  [4, 5, 6],
  [7, 8, 9]
]
print('Brute: ',brute(mat2))
print("Optimal: ",optimal(mat2))

File: Striver_C++_Problems/Arrays/P027.cpp
// P027.cpp - Rotate N x N matrix

#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> brute(const vector<vector<int>>& grid) {
    int n = grid.size();
    vector<vector<int>> res(n, vector<int>(n, 0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            res[j][n - 1 - i] = grid[i][j];
        }
    }
    return res;
}

vector<vector<int>> optimal(vector<vector<int>> grid) {
    int n = grid.size();
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            swap(grid[i][j], grid[j][i]);
        }
    }
    for (int i = 0; i < n; i++) {
        reverse(grid[i].begin(), grid[i].end());
    }
    return grid;
}

void printMatrix(const vector<vector<int>>& m) {
    cout << "[\n";
    for (auto &row : m) {
        cout << "  [";
        for (size_t j = 0; j < row.size(); j++) {
            cout << row[j];
            if (j + 1 < row.size()) cout << ", ";
        }
        cout << "]\n";
    }
    cout << "]\n";
}

int main() {
    vector<vector<int>> mat2 = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    };

    auto b = brute(mat2);
    auto o = optimal(mat2);

    cout << "Brute: \n";
    printMatrix(b);
    cout << "Optimal: \n";
    printMatrix(o);
    return 0;
}
